File: 문제풀이/2023/032723 백준 특별한 서빙 27896 G4.js
class MaxHeap {
  constructor() {
    this.heap = [];
  }
  hPush(value) {
    this.heap.push(value);
    this.heapifyUp(this.heap.length - 1);
  }
  hPop() {
    if (this.heap.length === 0) throw new Error('Cannot Pop Empty Heap');
    if (this.heap.length === 1) return this.heap.pop();

    const maxVal = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapifyDown(0);
    return maxVal;
  }
  heapifyUp(idx) {
    if (idx === 0) return;

    const parentIdx = Math.floor((idx - 1) / 2);
    if (this.heap[idx] > this.heap[parentIdx]) {
      [this.heap[parentIdx], this.heap[idx]] = [this.heap[idx], this.heap[parentIdx]];
      this.heapifyUp(parentIdx);
    }
  }
  heapifyDown(idx) {
    const leftIdx = idx * 2 + 1;
    const rightIdx = idx * 2 + 2;

    const left = this.heap[leftIdx] ? this.heap[leftIdx] : -Infinity;
    const right = this.heap[rightIdx] ? this.heap[rightIdx] : -Infinity;
    const nextIdx = left > right ? leftIdx : rightIdx;

    if (this.heap[idx] < this.heap[nextIdx]) {
      [this.heap[idx], this.heap[nextIdx]] = [this.heap[nextIdx], this.heap[idx]];
      this.heapifyDown(nextIdx);
    }
  }
}

const input = (() => {
  const { readFileSync } = require('fs');
  const line = readFileSync('/dev/stdin', 'utf8').toString().split('\n').reverse();
  return () => {
    return line.pop();
  };
})();
(function main() {
  try {
    const [N, M] = input().split(' ').map(Number);
    const sadIndexes = input().split(' ').map(Number);

    const answer = solution(N, M, sadIndexes);
    console.log(answer);
    return 1;
  } catch (err) {
    return 0;
  }
})();
function solution(N, M, sads) {
  const eaten = new MaxHeap();
  let cnt = 0,
    curSad = 0;

  for (const sad of sads) {
    curSad += sad;
    eaten.hPush(sad);

    if (curSad >= M) {
      cnt++;
      curSad -= 2 * eaten.hPop();
    }
  }

  return cnt;
}

File: 문제풀이/2023/032723 백준 특별한 서빙 27896 G4.py
from heapq import heappop, heappush
from sys import argv, maxsize as INF, stdin
input = stdin.readline
A, B = 'A', 'B'


class MaxHeap(list):
    def hPush(self, x):
        heappush(self, -x)

    def hPop(self):
        return -heappop(self)


def solution(N: int, M: int, sads: list[int]) -> int:
    heap = MaxHeap()
    curSad = 0
    cnt = 0

    for sad in sads:
        curSad += sad
        heap.hPush(sad)

        while curSad >= M:
            cnt += 1
            curSad -= 2*heap.hPop()

    return cnt


def main(*args: list[str]) -> int:
    N, M = map(int, input().split())
    sads = list(map(int, input().split()))

    answer = solution(N, M, sads)
    print(answer)
    return 1


if __name__ == '__main__':
    main(*argv)
